### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by emashliles**
@gemcfadyen @jsuchy

### CODE REVIEW

File: test/TurnUITests.java [hunk 1]
```diff
@@ -2,29 +2,99 @@
...
+        TurnUI turns = new TurnUI(printer, System.out, in);
+
+        turns.takeTurn(board, game);
+        turns.takeTurn(board, game);
+
+        assertEquals(board.markerAt(1), "O");
+    }
+
+    @Test
+    public void asksForInputAgainIfInputIsText() {
```
Comment (jsuchy) [inline test/TurnUITests.java:40]: Could you make these tests smaller? There is a ton of setup for each of these cases, each of which test the validity of the input. Why do you need to test this through the UI? I would want exactly one test that verifies that the user is asked again if the input in invalid, not one test for each possible type of invalid input. You still need those tests, but they shouldn't go through the UI.

File: src/com/TTT/TurnUI.java [hunk 2]
```diff
@@ -1,14 +1,47 @@
...
+        out.print(text);
+        String selection = scanner.nextLine();
+        return parseSelection(selection);
+    }
+
+    private int parseSelection(String selection) {
+        try{
+            return Integer.parseInt(selection) - 1;
+        }
+        catch(Exception e){
```
Comment (jsuchy) [inline src/com/TTT/TurnUI.java:43]: I'd avoid catching a type so high up the inheritance tree. Can you handle the specific exception that you expect here?

File: src/com/TTT/TurnUI.java [hunk 3]
```diff
@@ -1,14 +1,47 @@
...
+        this.out = out;
+        scanner = new Scanner(in);
+        scanner.useDelimiter("\n");
+    }
+
+    public void takeTurn(Board board, Game game) {
+        int selectedSpace = getPlayerInput(board, inputPrompt);
+
+        while(selectedSpace < 0 || selectedSpace > (board.size() - 1) || board.isOccupied(selectedSpace))
+        {
```
Comment (jsuchy) [inline src/com/TTT/TurnUI.java:25]: I'm looking for consistency in formatting. K&R style would have the opening / close braces for class / method at the same indentation. I don't think you should drop the brace to the next line within a method. I would personally like to see the brace on the same line in all cases so there's no difference.

File: test/TurnUITests.java [hunk 4]
```diff
@@ -2,29 +2,99 @@
...
+import static org.junit.Assert.assertTrue;
 
 public class TurnUITests {
+
+    private Board board;
+    private Game game;
+    private BoardPrinter printer;
+
+    @Before
+    public void setUp() throws Exception {
```
Comment (jsuchy) [inline test/TurnUITests.java:22]: I'd suggest avoiding checked exceptions, so having throws Exception should be removed.

File: test/TurnUITests.java [hunk 5]
```diff
@@ -2,29 +2,99 @@
...
+    public void zeroIndexSelection() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("5").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        TurnUI turns = new TurnUI(printer, out, in);
+
+        turns.takeTurn(board, game);
+
+        assertEquals(board.isOccupied(4), true);
+
```
Comment (jsuchy) [inline test/TurnUITests.java:98]: Whitespace.

Lots of nitpicks in this PR. In general, looks pretty good. Testing validation logic via the UI is the big thing that stands out to me.

**Comment by jsuchy** [comment]
When working on a project, I'd expect each comment on the PR to be addressed (either with a conversation in the thread, or with a commit). I'd like to see this PR get that sort of attention.

File: src/com/TTT/Main.java [hunk 6]
```diff
@@ -12,14 +10,14 @@ private static void run() {
...
         Game game = new Game(board);
-        TurnUI turns = new TurnUI();
+        TurnUI turns = new TurnUI(printer, System.out, System.in);
 
-        takeTurns(board, printer, game, System.in, turns);
+        takeTurns(board, game, turns);
     }
 
-    public static void takeTurns(Board board, BoardPrinter printer, Game game, InputStream in, TurnUI turns) {
+    public static void takeTurns(Board board, Game game, TurnUI turns) {
```
Comment (gemcfadyen) [inline src/com/TTT/Main.java:18]: I guess this is still not under test? Might be worth thinking about where the game loop should live.

File: src/com/TTT/TurnUI.java [hunk 7]
```diff
@@ -1,14 +1,47 @@
...
-        Scanner sc = new Scanner(in);
-        String selection = sc.nextLine();
-        game.doTurn(Integer.parseInt(selection));
+        out.print(text);
+        String selection = scanner.nextLine();
+        return parseSelection(selection);
+    }
+
+    private int parseSelection(String selection) {
+        try{
```
Comment (gemcfadyen) [inline src/com/TTT/TurnUI.java:40]: To be consistent I would leave a space between the `y` and the `{`

File: test/TurnUITests.java [hunk 8]
```diff
@@ -2,29 +2,99 @@
...
+        turns.takeTurn(board, game);
+
+        assertEquals(outputStream.toString(), "Please enter a space number: Invalid input. Please enter a space number: ");
+    }
+
+    @Test
+    public void asksForInputAgainIfInputIsTooLarge() {
+        ByteArrayInputStream in = new ByteArrayInputStream(("9999999999999999999999999999\n7").getBytes());
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
Comment (gemcfadyen) [inline test/TurnUITests.java:54]: There is some duplication amongst the tests - could it be extracted into some small methods, which might help simplify the tests a little.

File: test/TurnUITests.java [hunk 9]
```diff
@@ -2,29 +2,99 @@
...
+        board.placeMarker(4, "X");
+        TurnUI turns = new TurnUI(printer, out, in);
+
+        turns.takeTurn(board, game);
 
+        assertEquals(outputStream.toString(), "Please enter a space number: Invalid input. Please enter a space number: ");
+    }
+
+    @Test
+    public void zeroIndexSelection() {
```
Comment (gemcfadyen) [inline test/TurnUITests.java:89]: Something to think about going forwards is that the name of the test is usually linked a little with the assertions. At first, it is tricky to find the link between the name of this name and the assertion. Maybe changing the assertion to be something along the lines of: `assertEquals(board.isOccupied(zeroIndexed(userInput)), true)` might help firm up this link.

File: src/com/TTT/Main.java (hunk 6 above)
Comment (emashliles) [reply src/com/TTT/Main.java:18]: Could I do this with a larger end to end test - e.g. plays the whole game?

Comment (gemcfadyen) [reply src/com/TTT/Main.java:18]: Yes you could implement an end to end test that plays an entire game which would you coverage.

**Pull request merged**